### PR TITLE
Move from SharedDHKey to PerUserKey

### PR DESCRIPTION
**Pull request by mlsteele**
SharedDHKey -> PerUserKey

This includes a bunch of renaming as well as substantive changes to sigchain loading, keyring, and provisioners. This should bring client back to parity with how it was before churn. Revocation of any sort is still busted the flag on and the tests for that are skipped. Which means prevs and generations above 1 are never used because revoking is the only way to roll the per-user-key.

This includes a change to kex protocol format so I will manually test kexing across version boundaries before merging this.

Hopefully this will partially unblock team-folk to run things that do not involve revoking keys.

Before merge:
- [x] manually test kex across versions
- [x] manually test a smattering of things with puk-support off

Not in this PR:
- [ ] revoke with puk on
- [ ] replace flag puk-enable with two: puk-support and puk-upgrade

### CODE REVIEW

File: go/libkb/per_user_key.go [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+	}
+	upakArg := NewLoadUserByUIDArg(ctx, s.G(), s.uid)
+	upakArg.LoginContext = lctx
+	upak, _, err := s.G().GetUPAKLoader().Load(upakArg)
+	return upak, err
+}
+
+func (s *PerUserKeyring) mergeLocked(m PerUserKeyMap) (err error) {
+	for k, v := range m {
+		s.generations[k] = v
```
Comment (mlsteele) [inline go/libkb/per_user_key.go]: This is not a deep copy. But it points to an object that should never be mutated.
```
type perUserKeyFull struct {
	seed   PerUserKeySeed
	sigKey *NaclSigningKeyPair
	encKey *NaclDHKeyPair
}
```

**Comment by keybase-ci-visdiff** [comment]
The commits 42bb50eb28f4...e4bed8f21439 introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Markdown-emoji** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1494452716128-42bb50eb28f4-e4bed8f21439-linux/Markdown-emoji-42bb50eb28f4.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1494452716128-42bb50eb28f4-e4bed8f21439-linux/Markdown-emoji-e4bed8f21439.png)  
   Markdown-emoji comparison
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits 42bb50eb28f4...bcf1389340e3 introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Markdown-emoji** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1494453133371-42bb50eb28f4-bcf1389340e3-linux/Markdown-emoji-42bb50eb28f4.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1494453133371-42bb50eb28f4-bcf1389340e3-linux/Markdown-emoji-bcf1389340e3.png)  
   Markdown-emoji comparison
</summary>

File: go/libkb/naclwrap.go [hunk 2]
```diff
@@ -793,6 +793,25 @@ func DeriveSymmetricKey(inKey NaclSecretBoxKey, reason EncryptionReason) (NaclSe
 	return outKey, nil
 }
 
+func DeriveNonReversed(inKey [32]byte, reason DeriveReason) (outKey [32]byte, err error) {
```
Comment (mlsteele) [inline go/libkb/naclwrap.go]: This is poorly named. This and `DeriveSymmetricKey` do basically the same job but in different ways.

File: go/engine/device_keygen.go [hunk 3]
```diff
@@ -129,29 +125,33 @@ func (e *DeviceKeygen) Push(ctx *Context, pargs *DeviceKeygenPushArgs) error {
...
-	if e.G().Env.GetEnableSharedDH() {
-		e.G().Log.CDebugf(ctx.NetContext, "DeviceKeygen#Push SDH:%v", e.G().Env.GetEnableSharedDH())
+	if e.G().Env.GetSupportPerUserKey() {
+		e.G().Log.CDebugf(ctx.NetContext, "DeviceKeygen#Push PUK:%v", e.G().Env.GetSupportPerUserKey())
 	}
 
-	var sdhBoxes = []keybase1.SharedDHSecretKeyBox{}
-	if e.G().Env.GetEnableSharedDH() && e.args.IsEldest {
+	var pukBoxes = []keybase1.PerUserKeyBox{}
+	if e.G().Env.GetSupportPerUserKey() && e.args.IsEldest {
```
Comment (oconnor663) [inline go/engine/device_keygen.go]: `GetSupportPerUserKey` is hardcoded false right now, right?

Comment (mlsteele) [inline go/engine/device_keygen.go]: It is forced off for non-dev run-modes [here](https://github.com/keybase/client/pull/6939/files#diff-3ec785d6e391d57369aa99772ca02469R656). In dev it defaults to off and can be enabled with an env var or flag.

File: go/engine/device_keygen.go [hunk 4]
```diff
@@ -412,13 +401,58 @@ func (e *DeviceKeygen) prepareSDHBoxesFromPaperkey(ctx *Context) ([]keybase1.Sha
...
+
+	publicKeysEntry := make(libkb.JSONPayload)
+	publicKeysEntry["signing"] = pukSigKey.GetKID().String()
+	publicKeysEntry["encryption"] = pukEncKey.GetKID().String()
+
+	res := make(libkb.JSONPayload)
+	res["sig"] = sig
+	res["signing_kid"] = signer.GetKID().String()
+	res["type"] = libkb.LinkTypePerUserKey
+	res["public_keys"] = publicKeysEntry
```
Comment (oconnor663) [inline go/engine/device_keygen.go]: @maxtaco just checking, is this a legacy PGP-era sort of thing, where we're keeping up with it to make it easier to share server code?

Comment (maxtaco) [inline go/engine/device_keygen.go]: YES!

File: go/libkb/delegatekeyaggregator.go [hunk 5]
```diff
@@ -11,14 +11,19 @@ import (
 
 // DelegatorAggregator manages delegating multiple keys in one post to the server
 
+// When run produces a map which goes into the 'key/multi' 'sigs' list.
+type AggSigProducer func() (JSONPayload, error)
+
 // Run posts an array of delegations to the server. Keeping this simple as we don't need any state (yet)
-func DelegatorAggregator(lctx LoginContext, ds []Delegator, sdhBoxes []keybase1.SharedDHSecretKeyBox) (err error) {
+// `extra` is optinal and adds an extra sig, produced by something other than a Delegator, after the others.
```
Comment (oconnor663) [inline go/libkb/delegatekeyaggregator.go]: spelling of "optional"

File: go/libkb/delegatekeyaggregator.go [hunk 6]
```diff
@@ -56,3 +66,28 @@ func DelegatorAggregator(lctx LoginContext, ds []Delegator, sdhBoxes []keybase1.
...
+	if len(pukBoxes) > 0 {
+		section["boxes"] = pukBoxes
+		section["generation"] = pukBoxes[0].Generation
+	}
+	if pukPrev != nil {
+		section["secretbox"] = pukPrev
+	}
+	if len(section) > 0 {
+		serverArg["per_user_key"] = section
+	}
```
Comment (oconnor663) [inline go/libkb/delegatekeyaggregator.go]: Is there any case where we expect any of these `if` conditions to be false?

Comment (mlsteele) [inline go/libkb/delegatekeyaggregator.go]: Just the second one: no prev when adding a new device. The first and third are basically the same and just make it so this method can be called as no-op. But I don't like that anymore, so I'll change it to require boxes.

File: go/libkb/kbsig.go [hunk 7]
```diff
@@ -614,3 +612,35 @@ type SigMultiItem struct {
 	SigInner   string `json:"sig_inner"`
 	TeamID     string `json:"team_id"`
 }
+
+// PerUserKeyProof creates a proof introducing a new per-user-key generation.
+// `signingKey` is the key signing in this new key. Not to be confused with the derived per-user-key signing key.
+// `reverseSig` can be nil. A nil `reverseSig` can be used to produce the inner of the reverse sig.
+func PerUserKeyProof(me *User, pukSigKID keybase1.KID, pukEncKID keybase1.KID, generation keybase1.PerUserKeyGeneration,
+	signingKey GenericKey, reverseSig *string) (*jsonw.Wrapper, error) {
```
Comment (oconnor663) [inline go/libkb/kbsig.go]: Maybe split the args here up into their own lines? I don't know if we've been consistent about this thought. I'm kind of surprised gofmt doesn't have a strong opinion about it.

File: go/libkb/keyfamily.go [hunk 8]
```diff
@@ -579,14 +586,24 @@ func (cki *ComputedKeyInfos) Delegate(kid keybase1.KID, tm *KeybaseTime, sigid k
...
+		SigKID: s.sigKID,
+		EncKID: s.encKID,
+	}
+	if entry.Gen <= 0 {
+		return fmt.Errorf("invalid per-user-key generation %v", entry.Gen)
+	}
+	if entry.SigKID.IsNil() {
+		return errors.New("nil per-user-key sig kid")
+	}
+	if entry.EncKID.IsNil() {
```
Comment (oconnor663) [inline go/libkb/keyfamily.go]: Should the chain link have failed to parse earlier, if it's going to fail any of these checks?

File: go/libkb/naclwrap.go [hunk 9]
```diff
@@ -860,3 +879,13 @@ func (k NaclDHKeyPair) Decrypt(nei *NaclEncryptionInfo) (plaintext []byte, sende
...
+
+func GeneratePerUserKeySeed() (PerUserKeySeed, error) {
+	var seed [PerUserKeySeedSize]byte
+	if nRead, err := rand.Read(seed[:]); err != nil {
+		return seed, err
+	} else if nRead != len(seed) {
+		return seed, fmt.Errorf("Short random read: %d", nRead)
+	}
+	return seed, nil
+}
```
Comment (oconnor663) [inline go/libkb/naclwrap.go]: I feel like we have a few versions of this floating around. Would it be simpler to do `PerUserKeySeed(MakeByte32(RandBytes(32)))`?

File: go/libkb/per_user_key.go [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+func (s *PerUserKeySeed) DeriveSigningKey() (*NaclSigningKeyPair, error) {
+	derived, err := DeriveNonReversed(*s, DeriveReasonPUKSigning)
+	if err != nil {
+		return nil, err
+	}
+	res, err := MakeNaclSigningKeyPairFromSecret(derived)
+	return &res, err
+}
+
+func (s *PerUserKeySeed) DeriveEncryptionKey() (*NaclDHKeyPair, error) {
```
Comment (oconnor663) [inline go/libkb/per_user_key.go]: Might make sense to call this a DH key or something, because there might also be times we derive symmetric encryption keys? (Besides the prev key...key :) ...which has its own method below.)

File: go/libkb/per_user_key.go [hunk 11]
```diff
@@ -0,0 +1,598 @@
...
+// derivePrevKey derives the symmetric key used to secretbox the previous generation seed.
+func (s *PerUserKeySeed) derivePrevKey() (res NaclSecretBoxKey, err error) {
+	derived, err := DeriveNonReversed(*s, DeriveReasonPUKPrev)
+	if err != nil {
+		return res, err
+	}
+	return NaclSecretBoxKey(derived), err
+}
+
+func (s *PerUserKeySeed) IsBlank() bool {
```
Comment (oconnor663) [inline go/libkb/per_user_key.go]: Would it be safer to pass the seed around as a pointer, so that the default value is less likely to make it into any actual computation? (I'm not sure it matters, since slicing works on nil? Just thinking out loud.)

File: go/libkb/keyfamily.go (hunk 8 above)
Comment (mlsteele) [inline go/libkb/keyfamily.go]: This is the only check I can find. For subkey links I can't find a check either..

File: go/libkb/per_user_key.go [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+
+	var nonce [NaclDHNonceSize]byte
+	if nRead, err := rand.Read(nonce[:]); err != nil {
+		return "", err
+	} else if nRead != NaclDHNonceSize {
+		return "", fmt.Errorf("Short random read: %d", nRead)
+	}
+
+	// secretbox
+	var symmetricKey2 [NaclSecretBoxKeySize]byte = symmetricKey
```
Comment (oconnor663) [inline go/libkb/per_user_key.go]: Can this be done with just a cast, without needing the temporary?

File: go/libkb/per_user_key_test.go [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+func TestNewPerUserKeySecretBox2(t *testing.T) {
+	// Test opening a canned one
+
+	expectedSeed, err := hex.DecodeString("7c13fbe503aa2f79970e43eeb832820cfbf67ee4298ae5345ffe19cc1048bdee")
+	require.NoError(t, err)
+	key1, err := hex.DecodeString("29e48abe7de0da83023ee5741c2a8959624325f2e6b8ed2c96caf676d05d7cda")
+	key := MakeByte32(key1)
+	require.NoError(t, err)
+
+	sb := "kwG4BAQEAAAAAAAAAAAAAAAAAAAAAAAAAAAA2gAwWiGRZRwucAYHVQ4xJAX9Zn7F5WfnjOmjBd3n\nxkpS8xUFQhjmQRXAfUadTNqY5TU7\n"
```
Comment (oconnor663) [inline go/libkb/per_user_key_test.go]: Is that the nonce screaming? :-D

File: go/libkb/per_user_key.go [hunk 14]
```diff
@@ -0,0 +1,598 @@
...
+type PerUserKeyMap map[keybase1.PerUserKeyGeneration]perUserKeyFull
+
+// PerUserKeyring holds on to all versions of the per user key.
+// Generation=0 should be nil, but all others should be present.
+type PerUserKeyring struct {
+	Contextified
+	sync.Mutex
+	uid         keybase1.UID
+	generations PerUserKeyMap
+}
```
Comment (oconnor663) [inline go/libkb/per_user_key.go]: Not part of this PR, but what are our intentions for disk caching these keys? You mentioned @maxtaco was skeptical about it?

File: go/libkb/util.go [hunk 15]
```diff
@@ -600,3 +600,13 @@ func MakeByte32(a []byte) [32]byte {
...
+
+func MakeByte32Soft(a []byte) ([32]byte, error) {
+	const n = 32
+	var b [n]byte
+	if len(a) != n {
+		return b, fmt.Errorf("MakeByte expected len %v but got %v slice", n, len(a))
+	}
+	copy(b[:], a)
+	return b, nil
+}
```
Comment (oconnor663) [inline go/libkb/util.go]: Ah sorry, the function above used to be this, and I swapped it out because we didn't have any use cases yet. That said, in at least one of the cases where you're calling this, you are actually checking the length beforehand. Could drop that check maybe?

File: go/libkb/per_user_key.go (hunk 11 above)
Comment (mlsteele) [inline go/libkb/per_user_key.go]: I'm not sure that would be safer because then you have 3 possible states: nil, blank, good. I think checking before use is good. Open to changing it, but not sure about the gains.

File: go/libkb/per_user_key_test.go (hunk 13 above)
Comment (mlsteele) [inline go/libkb/per_user_key_test.go]: `crypto.random.SomeA's(24)`

File: go/libkb/util.go (hunk 15 above)
Comment (mlsteele) [inline go/libkb/util.go]: 👍  switched that one to `MakeByte32`

File: go/libkb/per_user_key.go (hunk 12 above)
Comment (mlsteele) [inline go/libkb/per_user_key.go]: Yes!
